src/mem/mem_scheduler.cc

```diff
@@ -0,0 +1,435 @@
+
+void
+MemScheduler::MemSidePort::recvRangeChange()
+{
+    owner->sendRangeChange();
+}
+
+bool
+MemScheduler::handleRequest(CPUSidePort *port, PacketPtr pkt)
+{
+    panic_if(!(pkt->isRead() || pkt->isWrite()),
+             "Should only see read and writes at memory controller\n");
+
+    uint32_t requestorId = pkt->req->requestorId();
+
+    std::unordered_map<RequestorID, CPUSidePort*>::const_iterator ret = routingTable.find(requestorId);
+    if (ret == routingTable.end())
+        routingTable[requestorId] = port;
+
+    std::unordered_map<RequestorID, bool>::const_iterator rit = readBlocked.find(requestorId);
+    if (rit == readBlocked.end())
+        readBlocked[requestorId] = false;
+
+    std::unordered_map<RequestorID, bool>::const_iterator wit = writeBlocked.find(requestorId);
+    if (wit == writeBlocked.end())
+        writeBlocked[requestorId] = false;
+
+    if (pkt->isRead() && readBlocked[requestorId])
+        return false;
+
+    if (pkt->isWrite() && writeBlocked[requestorId])
+        return false;
+
+    DPRINTF(MemScheduler, "Got request for addr %#x for requestorId: %d\n", pkt->getAddr(), requestorId);
+
+    if (pkt->isRead()){
+        DPRINTF(MemScheduler, "handleRequest: pushing read request, readQueue[%d].size(): %d\n", requestorId, readQueues[requestorId].size());
+        readQueues[requestorId].push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: pushed read request, readQueue[%d].size(): %d\n", requestorId, readQueues[requestorId].size());
+        if(readQueues[requestorId].size() == readBufferSize)
+            readBlocked[requestorId] = true;
+    }
+    if (pkt->isWrite()){
+        DPRINTF(MemScheduler, "handleRequest: pushing write request, writeQueue[%d].size(): %d\n", requestorId, writeQueues[requestorId].size());
+        writeQueues[requestorId].push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: pushed write request, writeQueue[%d].size(): %d\n", requestorId, writeQueues[requestorId].size());
+        if(writeQueues[requestorId].size() == writeBufferSize)
+            writeBlocked[requestorId] = true;
+    }
+    if (!nextReqEvent.scheduled()){
+        if (pkt->isRead())
+            currentReadEntry = readQueues.find(requestorId);
+        if (pkt->isWrite())
+            currentWriteEntry = writeQueues.find(requestorId);
+        DPRINTF(MemScheduler, "Scheduling nextReqEvent in handleRequest\n");
+        schedule(nextReqEvent, curTick());
+    }
+    return true;
+}
+
+MemScheduler::MemSidePort*
+MemScheduler::findMemoryPort(PacketPtr pkt){
+    const Addr base_addr = pkt->getAddr();
+    for (auto memPort : memPorts)
+        for (auto addr_range : memPort->getAddrRanges())
+            if (addr_range.contains(base_addr)){
+                return memPort;
+            }
+    return NULL;
+}
+
+void
+MemScheduler::processNextReqEvent(){
+    std::unordered_map<RequestorID, std::queue<PacketPtr> >::iterator \
+                                        initialEntry = currentReadEntry;
```

Comment:
`auto` is your friend!
```suggestion
    auto initialEntry = currentReadEntry;
```